python/ray/data/impl/pipeline_executor.py

```diff
@@ -60,6 +49,32 @@ def __init__(self, pipeline: "DatasetPipeline[T]"):
         else:
             self._bars = None
 
+    def __del__(self):
+        for f in self._stages:
+            if f is not None:
+                f.cancel()
+        self._pool.shutdown(wait=False)
+
+        # Signal to all remaining threads to shut down.
+        with progress_bar._canceled_threads_lock:
+            for t in self._pool._threads:
+                if t.is_alive():
+                    progress_bar._canceled_threads.add(t)
+
+        # Wait for 1s for all threads to shut down.
+        start = time.time()
+        while time.time() - start < 1:
+            self._pool.shutdown(wait=False)
```

Comment:
What is this repeated shutdown call supposed to do? Is this just for repeatedly sending a wake-up signal to the worker threads? https://github.com/python/cpython/blob/main/Lib/concurrent/futures/thread.py#L232

Reply:
It's just to give some time for the pool to shut down. It's best to clean up all the threads, but if we can't, we shouldn't hang.